Add config flow for rachio

bdraco opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Config flow for rachio

Also discoverable via homekit


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/12354

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Do not fail when a user has a controller with shared access on their account

bdraco committed:
Add config flow for rachio

Also discoverable via homekit

balloob commented on homeassistant/components/rachio/.translations/en.json:
```diff
@@ -0,0 +1,24 @@
+        "data": {
+          "api_key": "The API key for the Rachio account.",
+          "hass_url_override": "If your instance is unaware of its actual web location (base_url).",
+          "manual_run_mins": "For how long, in minutes, to turn on a station when the switch is enabled."
```
Let's move this to an option flow. Normal setup should focus on the things necessary to get it in Home Assistant. Nothing more.

bdraco replied:
moved to option flow

balloob commented on homeassistant/components/rachio/binary_sensor.py:
```diff
@@ -70,6 +76,22 @@ def _poll_update(self, data=None) -> bool:
+                    DOMAIN_RACHIO,
+                    self._controller.controller_id,
+                    self._controller.serial_number,
+                    self._controller.mac_address,
```
Let's pick 1. There is no need use 3 unique values.

balloob replied:
Also please return the mac address as a `connections` with `device_registry.CONNECTION_NETWORK_MAC`

bdraco replied:
controller_id might change if rachio changes something but the serial number is fixed in the real world so will go with that

bdraco replied:
added mac to connections

balloob commented on homeassistant/components/rachio/switch.py:
```diff
@@ -33,22 +37,31 @@
-    add_entities(devices)
+    for controller in hass.data[DOMAIN_RACHIO][config_entry.entry_id].controllers:
+        devices.append(
+            await hass.async_add_executor_job(RachioStandbySwitch, hass, controller)
```
We should avoid jumping often into the executor. Better to create a sync function that does this all and pass that to the executor.

bdraco replied:
fixed

balloob commented on homeassistant/components/rachio/switch.py:
```diff
@@ -160,6 +189,7 @@ def __init__(self, hass, controller, data, manual_run_time):
         self._zone_number = data[KEY_ZONE_NUMBER]
         self._zone_enabled = data[KEY_ENABLED]
         self._manual_run_time = manual_run_time
+        self._entity_picture = data[KEY_IMAGE_URL] if KEY_IMAGE_URL in data else None
```
```suggestion
        self._entity_picture = data.get(KEY_IMAGE_URL)
```

bdraco committed:
Update homeassistant/components/rachio/switch.py

Co-Authored-By: Paulus Schoutsen <paulus@home-assistant.io>

balloob commented on homeassistant/components/rachio/__init__.py:
```diff
@@ -102,28 +101,62 @@
     # Get the URL of this server
-    custom_url = config[DOMAIN].get(CONF_CUSTOM_URL)
+    custom_url = config.get(CONF_CUSTOM_URL)
     hass_url = hass.config.api.base_url if custom_url is None else custom_url
```
What's going on here? I know that it's existing code but this seems odd. This seems like working around the fact that people don't have set up their base url correctly ? We should use `async_get_external_url` from the network helper.

bdraco replied:
Removed from config flow.  Now yaml only.  I will likely remove it when I do the webhooks conversion assuming.  Does it look like the cloud hooks solve this anyways?

bdraco replied:
Leaving this for now for a followup

balloob replied:
Yeah cloud hooks will bypass base url completely.

balloob commented on homeassistant/components/rachio/__init__.py:
```diff
@@ -102,28 +101,62 @@
-    except AssertionError as error:
+        person = await hass.async_add_executor_job(RachioPerson, hass, rachio, config)
+    # Yes we really do get all these exceptions (hopefully rachiopy switches to requests)
+    except (http.client.HTTPException, ssl.SSLError, OSError, AssertionError) as error:
```
You can consider putting these exceptions in a reusable variable. 

btw, assertion error 🙄

balloob commented on homeassistant/components/rachio/__init__.py:
```diff
@@ -285,15 +356,18 @@ class RachioWebhookView(HomeAssistantView):
     url = WEBHOOK_PATH
```
This should be rewritten to use our webhooks implementation in a future PR.

balloob replied:
Actually, having this fixed url being bound to a single entry means that right now you are still limited to 1 rachio account until it starts using our built-in webhooks

bdraco replied:
I'm thinking the interim step is to append the config entry_id to the url and then switch this to the way netatmo does it in a new PR (if that is the best example)

bdraco replied:
I added the entry_id to the URL so they have a unique one.  I used two different accounts to test with and ensured that the live updates came back to the separate endpoints.

bdraco replied:
Converted to cloud hooks here
https://github.com/home-assistant/core/pull/33724

balloob commented on homeassistant/components/rachio/config_flow.py:
```diff
@@ -0,0 +1,104 @@
+    {
+        vol.Required(CONF_API_KEY): str,
+        vol.Optional(CONF_CUSTOM_URL): str,
+        vol.Optional(CONF_MANUAL_RUN_MINS, default=DEFAULT_MANUAL_RUN_MINS): int,
```
Remove this from what we ask users during setup.

bdraco commented:
I'm blocked for testing at the moment because I hit the api limit.  I also realized we fire off self._controller.current_schedule for each zone that gets added.  For testing, I added an account with 64 zones.  Looks like I'm going to have to fix that before I can continue

bdraco commented:
Last commit seemed to take care of it.  the startup is MUCH MUCH faster :)

bdraco committed:
Split setting the default run time to an options flow

Ensue the run time coming from yaml gets imported into the option flow

Only get the schedule once at setup instead of each zone (was hitting rate limits)

Add the config entry id to the end of the webhook so there is a unique hook per config entry

Breakout the slew of exceptions rachiopy can throw into RachioAPIExceptions

Remove the base url override as an option for the config flow

Switch identifer for device_info to serial number

Add connections to device_info (mac address)

bdraco commented:
There is an existing bug where if you add a controller to two different home assistants only the last one gets the notifications.  Confirmed not a regression.  I'll try to fix that when I convert the webhooks in a new PR

bdraco commented:
Tested on two different installs
Tested an account with 7 controllers.  1 shared
Tested an account with 1 controller. 0 shared
Verified webhooks work with multiple controllers.

bdraco committed:
rename to make pylint happy

bdraco committed:
Fix import of custom_url

bdraco commented:
Retested after last commit
Added yaml for import from yaml with 27m watering time
Started watering and it did 27m
Verified I could change it to 12m
Started watering and it did 12m
Created new integration for different rachio
Started watering and it did 10m (default)
Verified I could change it to 3m
Started watering and it did 3m

bdraco commented:
Lots of restarts and no more rate limiting since I fixed the current_schedule api being called per zone

balloob commented on homeassistant/components/rachio/binary_sensor.py:
```diff
@@ -70,6 +78,18 @@ def _poll_update(self, data=None) -> bool:
+        return {
+            "identifiers": {(DOMAIN_RACHIO, self._controller.serial_number,)},
+            "connections": {
+                (device_registry.CONNECTION_NETWORK_MAC, self._controller.mac_address,)
```
This is cool, because now other integrations can add data to this device, like Unifi can tell you if it's online/offline.

balloob commented on homeassistant/components/rachio/manifest.json:
```diff
@@ -2,7 +2,17 @@
+  "dependencies": [
+    "http"
+  ],
+  "codeowners": [],
```
You want to claim it?

bdraco replied:
Might as well as I have quite a few of these

bdraco replied:
done

bdraco replied:
well on the wrong branch.. tired.. fixing

bdraco replied:
done for real

bdraco committed:
claim rachio

MartinHjelmare reviewed (commented):
Nice! Some comments for future PRs.

bdraco commented on homeassistant/components/rachio/__init__.py:
```diff
@@ -102,28 +100,69 @@
+
+    # CONF_MANUAL_RUN_MINS can only come from a yaml import
+    if not options.get(CONF_MANUAL_RUN_MINS) and config.get(CONF_MANUAL_RUN_MINS):
+        options[CONF_MANUAL_RUN_MINS] = config[CONF_MANUAL_RUN_MINS]
```
Thanks for the review .  I’ll take care of these In the next round

balloob replied:
Options are frozen, I'm surprised this works.

bdraco commented on homeassistant/components/rachio/__init__.py:
```diff
@@ -282,18 +359,22 @@ class RachioWebhookView(HomeAssistantView):
+        self._entry_id = entry_id
+        self.url = webhook_url
+        self.name = webhook_url[1:].replace("/", ":")
+        _LOGGER.debug("Created webhook at url: %s, with name %s", self.url, self.name)
```
will fix the message

bdraco commented on homeassistant/components/rachio/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        _LOGGER.debug("async_step_user: %s", user_input)
```
Yup, thats bad. fixing

bdraco commented on homeassistant/components/rachio/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+            try:
+                info = await validate_input(self.hass, user_input)
+                await self.async_set_unique_id(user_input[CONF_API_KEY])
+                return self.async_create_entry(title=info["title"], data=user_input)
```
Yes. done

bdraco commented:
@MartinHjelmare review items addressed in https://github.com/home-assistant/core/pull/32814
